internal/services/kaas/kaas_resource.go

```diff
@@ -564,91 +563,95 @@ func (r *kaasResource) Update(ctx context.Context, req resource.UpdateRequest, r
 		return
 	}
 
-	// Update API call logic
+	input := r.prepareUpdateInput(state, data, chosenPackState.Id)
+
+	if err := r.performAPIUpdate(input); err != nil {
+		resp.Diagnostics.AddError("Error when updating KaaS", err.Error())
+		return
+	}
+
+	kaasObject, err := r.waitUntilActive(ctx, input, input.Id)
+	if err != nil || kaasObject == nil {
+		resp.Diagnostics.AddError("Error waiting for KaaS activation", err.Error())
+		return
+	}
+
+	r.fetchAndSetKubeconfig(&data, input, resp)
+
+	data.fill(kaasObject)
+
+	if data.Apiserver != nil {
+		r.handleApiserverConfig(ctx, &data, input, resp)
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *kaasResource) prepareUpdateInput(state, data KaasModel, packID int) *kaas.Kaas {
 	input := &kaas.Kaas{
 		Project: kaas.KaasProject{
 			PublicCloudId: int(data.PublicCloudId.ValueInt64()),
 			ProjectId:     int(data.PublicCloudProjectId.ValueInt64()),
 		},
 		Id:                int(state.Id.ValueInt64()),
 		Name:              data.Name.ValueString(),
-		PackId:            chosenPackState.Id,
+		PackId:            packID,
 		Region:            state.Region.ValueString(),
-		KubernetesVersion: state.KubernetesVersion.ValueString(),
+		KubernetesVersion: data.KubernetesVersion.ValueString(),
 	}
 
-	_, err = r.client.Kaas.UpdateKaas(input)
-	if err != nil {
-		resp.Diagnostics.AddError(
-			"Error when updating KaaS",
-			err.Error(),
-		)
-		return
+	if state.KubernetesVersion.ValueString() == data.KubernetesVersion.ValueString() {
+		input.KubernetesVersion = ""
 	}
 
-	kaasObject, err := r.waitUntilActive(ctx, input, input.Id)
-	if err != nil {
-		resp.Diagnostics.AddError(
-			"Error when getting KaaS",
-			err.Error(),
-		)
-		return
-	}
+	return input
+}
 
-	if kaasObject == nil {
-		return
-	}
+func (r *kaasResource) performAPIUpdate(input *kaas.Kaas) error {
+	_, err := r.client.Kaas.UpdateKaas(input)
+	return err
+}
 
-	// Wait for kubeconfig to be available
+func (r *kaasResource) fetchAndSetKubeconfig(data *KaasModel, input *kaas.Kaas, resp *resource.UpdateResponse) {
```

Comment:
This is used once in the file, however there's multiple calls to `r.client.Kaas.GetKubeconfig`, we should probably use it there too in order to remove duplication ?